### PR TITLE
Add "sender" and "receiver" stats.

### PR DESCRIPTION
**Pull request by jan-ivar**
Fix for https://github.com/w3c/webrtc-stats/issues/231 (includes/on-top-of patch from https://github.com/w3c/webrtc-stats/pull/272).

### CODE REVIEW

**Comment by vr000m** [comment]
As per the discussion at TPAC17, I am opposed to removing the track stats. The decision was to keep the both the track and the sender and receiver stats for the forseeable future. If implementation experience suggests that the tracks can be removed, we can revisit the discussion at the time.

**Comment by jan-ivar** [comment]
@vr000m I'll update this PR to keep the track stats around. Just haven't had the time.

**Comment by alvestrand** [comment]
@jan-ivar ping?
This is probably the PR that is most likely to invalidate other PRs, so it would be nice to land it.

File: webrtc-stats.html [hunk 1]
```diff
@@ -472,15 +473,27 @@ <h3>
             </p>
           </dd>
           <dt>
-            <dfn><code>track</code></dfn>
+            <dfn><code>sender</code></dfn>
           </dt>
           <dd>
             <p>
-              Contains statistics related to a specific MediaStreamTrack and the corresponding
+              Contains statistics related to a specific sender and the corresponding
```
Comment (henbos) [inline webrtc-stats.html]: Here and elsewhere: prefer RTCRtpSender/RTCRtpReceiver. I realize this is a mix of moved text and new text though so we can do it separately if need be.

File: webrtc-stats.html [hunk 2]
```diff
@@ -472,15 +473,27 @@ <h3>
             </p>
           </dd>
           <dt>
-            <dfn><code>track</code></dfn>
+            <dfn><code>sender</code></dfn>
```
Comment (henbos) [inline webrtc-stats.html]: Should we have sender, receiver AND track? The sender/receiver referencing trackId.

In the sender case, you might expect counters to go up for the sender and for the current track, but as soon as you replaceTrack a different counter is used.

How useful are the track stats after we have sender/receiver stats? Which should be kept? @vr000m

**Comment by vr000m** [comment]
This should be to keep track, sender, and receiver. @jan-ivar could you update this PR?

**Comment by jan-ivar** [comment]
Working on it.

**Comment by alvestrand** [comment]
@jan-ivar happy new year! Still working on it?

**Comment by jan-ivar** [comment]
Incorporated feedback from @henbos in https://github.com/w3c/webrtc-stats/pull/272#discussion_r156664045.

Note: It's sufficient to merge this PR only rather than https://github.com/w3c/webrtc-stats/pull/272 (if https://github.com/w3c/webrtc-stats/pull/272 is merged I'll need to rebase again).

**Comment by alvestrand** [comment]
I noticed that since we've now added "objectDeleted", the only reason for having a class for track stats has disappeared - "detached" is now served by "objectDeleted" (at least until we decide to drop them both).

File: webrtc-stats.html [hunk 3]
```diff
@@ -1859,23 +1841,56 @@ <h2>
...
                   is the current decoding framerate.
                 </p>
               </dd>
+            </dl>
+          </section>
+        </div>
+      </section>
+      <section id="vsstats-dict*">
+        <h3>
+          <dfn>RTCVideoSenderStats</dfn> dictionary
```
Comment (henbos) [inline webrtc-stats.html]: We have these hierarchies (and similar for audio and receive):

`RTCMediaStreamTrackStats <- RTCVideoStreamTrackStats <- RTCVideoSenderStats <- RTCLocalVideoStreamTrackStats`

In other words

`Track <- [Audio/Video]Track <- [Audio/Video][Sender/Receiver] <- Local[Audio/Video]Track`

It gets a bit confusing about going back and forth between track language and sender/receiver language.

The way I see it, it's sender/receiver all the way down. The `RTCLocal[Audio/Video]StreamTrackStats` are the only special case where now we're not talking about a _sender_, now it changes to an _attachment_ of a track (it's not even a track-stat, since attaching the same track multiple times yields different attachment stats). Which really is a subview of the sender stats for the duration a track is attached.

Lacking a better word that can be used to refer to both a sender and a receiver than "SenderReceiver" (and avoiding confusion with `RTCRtpTransceiver`), I would name them:
```
RTCSenderReceiverStats (RTCMediaStreamTrackStats)
RTC[Audio/Video]SenderReceiverStats (RTC[Audio/Video]StreamTrackStats)
RTC[Audio/Video][Sender/Receiver]Stats (same name as before)
RTCSender[Audio/Video]TrackAttachmentStats (RTCLocal[Audio/Video]StreamTrackStats)
```

In other words

`SenderReceiver <- [Audio/Video]SenderReceiver <- [Audio/Video][Sender/Receiver] <- [Audio/Video]TrackAttachment`

But renames can be done in separate PRs.

**Comment by alvestrand** [comment]
I think we should merge this one, and iterate over a) naming and b) whether we need the track class at all.

**Comment by jan-ivar** [comment]
I've just removed the `detached` member for now. We can refactor away the dictionary entirely later perhaps.

File: webrtc-stats.html [hunk 4]
```diff
@@ -1859,23 +1841,56 @@ <h2>
...
               </dt>
               <dd>
                 <p>
-                  Only valid for local video. It represents the total number of frames captured for
-                  this MediaStreamTrack, before encoding. For example, if this track represents a
-                  camera this is the number of frames produced by the camera for this track, whose
-                  framerate could vary due to hardware limitations or environmental factors such as
-                  lighting conditions.
+                  Represents the total number of frames captured, before encoding,
+                  for this RTCRtpSender (or for this MediaStreamTrack, if
```
Comment (henbos) [inline webrtc-stats.html]: nit: Here and elsewhere: "...for this MediaStreamTrack _attachment_..."

**Comment by henbos** [comment]
Thanks so much for doing this @jan-ivar, this was a large one!!

File: webrtc-stats.html [hunk 5]
```diff
@@ -520,16 +522,40 @@ <h3>
...
-          <dt>
+           <dt>
             <dfn><code>track</code></dfn>
+           </dt>
+           <dd>
+             <p>
+              Contains statistics related to a specific MediaStreamTrack and the corresponding
+              media-level metrics. It is accessed by either
+              <code><a>RTCLocalVideoStreamTrackStats</a></code> or
+              <code><a>RTCLocalAudioStreamTrackStats</a></code>, both inherited from
```
[the remark this answers is not in the document]
Comment (jan-ivar) [reply webrtc-stats.html]: This doesn't remove "track" stats, so no breaking change (yet). The receiving end would presumably see both "track" and "receiver" with redundant info.

File: webrtc-stats.html (hunk 4 above)
Comment (jan-ivar) [reply webrtc-stats.html]: Right, though that was probably an existing issue? This PR mostly moves words around.

**Comment by alvestrand** [comment]
Editors meeting: Let's merge and fix naming etc. afterwards.
